### PR TITLE
Fuzzing updates: unit drivers

### PR DESCRIPTION
**Pull request by zcduthie**
A colleague and I have completed some research investigating fuzzing.
This pull request includes some of our findings for the basic unit fuzz drivers:

### 1. fuzz-chip-cert
- Add dictionary (use TLV dictionary described below)
- Add seed corpus (extracted from unit tests)
- Other:
  - Add coverage: additional calls to ChipCert methods
  - Remove ExtractSubjectDNFromX509Cert (and create new driver `fuzz-der-cert` to separate certificate fuzzing into appropriate binary formats)

Line Coverage Result: +366%
(for 3 target files in: src/credentials/CHIPCert**)

### 1a. fuzz-der-cert (NEW DRIVER)
(fuzzes certificates in DER binary format)
- Add dictionary (ASN1 tags?????)
- Add seed corpus (extracted from unit tests)

Line Coverage Result: 80.5% (NEW)
(for target file: src/credentials/CHIPCertFromX509.cpp)

### 2. fuzz-minmdns-packet-parsing:
- Add dictionary (mDNS constants extracted from src/lib/dnssd/minimal_mdns/core/Constants.h)
- Add seed corpus (manually generated by running Docker emulated lock-app)

Line Coverage Result: +3% (added final 'success' line: `return true`)
(for target file: src/lib/dnssd/minimal_mdns/Parser.cpp)

### 3. fuzz-tlv-reader
- Add dictionary (set of anonymous and context-specific tag control bytes)
- Add seed corpus (extracted from unit tests)
- Other:
  - Add coverage: additional method calls (including span + byte reads)

Line Coverage Result: +82%
(for 4 target files in: src/lib/core/CHIPTLV**)

### 4. fuzz-qr-setup-payload-parser (NEW DRIVER)
- Add dictionary (extracted from unit tests including base38 boundary values)
- Add seed corpus (extracted from unit tests)

Line Coverage Result: 88.2% (NEW)
(for target file: src/setup_payload/QRCodeSetupPayloadParser.cpp)

Note that the corpera in this PR have not been minimized yet (for ease of reference).

## Integration with OSS-Fuzz
Hopefully I've structured this pull request with default naming `xxx.dict` and `xxx-cert-corpus/` so that it's easily compatible with the OSS-Fuzz integration via something like:

```
# OSS-Fuzz projects/connectedhomeip/build.sh
# ...
# ...

for fuzzer in fuzz-chip-cert \
                fuzz-der-cert \
                fuzz-minmdns-packet-parsing \
                fuzz-tlv-reader \
                fuzz-qr-setup-payload-parser do
    cp integrations/fuzz/${fuzzer}.dict $OUT
    zip -j $OUT/${fuzzer}-cert-corpus.zip integrations/fuzz/${fuzzer}-corpus/*
done
```

Reference: [boringssl](https://github.com/google/oss-fuzz/blob/master/projects/boringssl/build.sh)

I can't verify the exact OSS-Fuzz script right now as OSS-Fuzz/connectedhomeip does not successfully run for me locally ([error attached](https://github.com/project-chip/connectedhomeip/files/12001489/oss-fuzz-connectedhomeip.log)). Has anyone got it running locally using `infra/helper.py build_fuzzers connectedhomeip`?

(I've also created a separate pull request for some more involved modifications to all-clusters-app fuzzing)

### CODE REVIEW

File: src/lib/core/tests/FuzzTlvReader.cpp [hunk 1]
```diff
@@ -4,10 +4,87 @@
...
+    {
+        aReader.GetType();
+    }
+    {
+        size_t size;
+        aReader.CountRemainingInContainer(&size);
+    }
+    {
+        bool v;
+        aReader.Get(v);
```
Comment (andy31415) [inline src/lib/core/tests/FuzzTlvReader.cpp:29]: Likly outside of scope, however I really wish getters would be `const`. A reader with non-const getters seems silly :/

Comment (bzbarsky-apple) [reply src/lib/core/tests/FuzzTlvReader.cpp:29]: Those getters can pretty radically change the state of the reader, though....

File: src/lib/core/TLVUtilities.cpp [hunk 2]
```diff
@@ -106,6 +106,35 @@ static CHIP_ERROR Iterate(TLVReader & aReader, size_t aDepth, IterateHandler aHa
...
+ *
+ *  @retval  The last value returned by @a aHandler, if different than #CHIP_NO_ERROR
+ *
+ */
+CHIP_ERROR Iterate(TLVReader & aReader, IterateHandlerMutable aHandler, void * aContext)
+{
+    const bool recurse = true;
+    CHIP_ERROR retval;
+
+    retval = Iterate(aReader, (IterateHandler) aHandler, aContext, recurse);
```
Comment (bzbarsky-apple) [inline src/lib/core/TLVUtilities.cpp:133]: This case is a bit suspect.  It might be better to just change the signature of IterateHandler to not have a const reader.

Either way, this core TLV change should be in a separate PR so it can get reviewed carefully.

Comment (zcduthie) [reply src/lib/core/TLVUtilities.cpp:133]: Good point. Especially given the context, I'll remove all non-const operations from TLV (and fuzzer) for now.

**Comment by zcduthie** [comment]
I've removed TLV operations that are not const (and associated changes) for now.

These can be re-added to the fuzzer later if desired either by modifying (safe) TLVReader getters to be const, or by modifying the iterator to not be const.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by andy31415 — DISMISSED** [review]
I would prefer to not check in the corpus - I believe the fuzzers can just run unattended and the corpus would be re-created (maybe not as efficiently).

Interested in alternate opinions though, but overall maybe lets create a PR without corpus and discuss corpus yes/no in a followup PR.

File: integrations/fuzz/fuzz-der-cert.dict
Comment (andy31415) [inline integrations/fuzz/fuzz-der-cert.dict:1]: Could we have some readme regarding usage? I am not sure what `.dict` files are and do not see references to them.

**Comment by andy31415** [comment]
> I would prefer to not check in the corpus - I believe the fuzzers can just run unattended and the corpus would be re-created (maybe not as efficiently).
> 
> Interested in alternate opinions though, but overall maybe lets create a PR without corpus and discuss corpus yes/no in a followup PR.

I beleive we also may need some of the description from the summary extracted into project documentation files for having people running fuzz tests.

There seems to be some limited docs in https://project-chip.github.io/connectedhomeip-doc/examples/all-clusters-app/linux/README.html#fuzzing-integration and https://project-chip.github.io/connectedhomeip-doc/guides/BUILDING.html#libfuzzer-unit-tests

File: integrations/fuzz/fuzz-der-cert.dict
Comment (zcduthie) [reply integrations/fuzz/fuzz-der-cert.dict:1]: Good idea. I'll add some extra information into the docs below.

Regarding seeds, are you happy to leave them checked in for now? Or perhaps I can 'pre-process' or minimise them to reduce the number of seeds (although this will mean they no longer mirror the crafted values from the Unit Tests)

Comment (zcduthie) [reply integrations/fuzz/fuzz-der-cert.dict:1]: @andy31415 I've added some sample commands and information text that hopefully gives a bit of an overview.
FWIW I've included links to google's fuzzing docs (https://github.com/google/fuzzing/blob/master/docs/glossary.md) which contains further information about seeds / dictionaries. Let me know if this is not suitable.

**Comment by andy31415** [comment]
PR not followed up in > 1 year  and has merge conflicts. Closing as stale.
